### PR TITLE
Check incoming transactions against current best tip ledger

### PR DESCRIPTION
**Pull request by enolan**
This should be the last thing that was causing us to keep invalid transactions
in the pool.

Check that pool transactions are valid against the current best tip on startup, when bootstrap finishes, when we change forks, and when they first come in. I moved the code in the `User_command` module defined in `coda_main.ml` into the one `user_command.ml` for simplicity's sake.

Added automated tests.

Checklist:

- [x] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [x] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

### CODE REVIEW

File: src/lib/transaction_pool/transaction_pool.ml [hunk 1]
```diff
@@ -20,10 +28,43 @@ module type Transition_frontier_intf = sig
...
   end
 
+  val best_tip : t -> Breadcrumb.t
+
   val best_tip_diff_pipe :
     t -> Extensions.Best_tip_diff.view Broadcast_pipe.Reader.t
 end
 
+module type User_command_intf = sig
+  type t [@@deriving sexp, bin_io]
```
Comment (bkase) [inline src/lib/transaction_pool/transaction_pool.ml]: See https://github.com/CodaProtocol/coda/blob/master/rfcs/0015-module-versioning.md , `bin_io` shouldn't be used directly on `type t` anymore.

Would be nice if we could just directly depend on `user_command` but I guess it's in coda_base?

Comment (bkase) [reply src/lib/transaction_pool/transaction_pool.ml]: ping @psteckler if you have any questions on how to stablify it

Comment (enolan) [reply src/lib/transaction_pool/transaction_pool.ml]: We can just depend on user command directly, but it's harder to test, hence the functoring. The `Make` functor doesn't take a user command parameter, but `Make0` does.

File: src/lib/transaction_pool/transaction_pool.ml [hunk 2]
```diff
@@ -88,23 +134,71 @@ struct
...
-    (* TODO We need to recheck validity here. Transactions from the old best
-       chain may not be valid on the new chain. *)
-    List.iter new_user_commands ~f:(fun tx -> remove_tx t tx) ;
+    let removed_set = User_command.Set.of_list removed_user_commands in
+    let added_set = User_command.Set.of_list new_user_commands in
+    let removed_but_not_added = User_command.Set.diff removed_set added_set in
+    (* Only considering commands that were added to the best tip and not removed
+       is an optimization, we have the invariant that a command in the pool is
+       never in the best tip. *)
+    let added_but_not_removed = User_command.Set.diff added_set removed_set in
```
[the remark this answers is not in the document]
Comment (enolan) [reply src/lib/transaction_pool/transaction_pool.ml]: it doesn't partition it though. it excludes transactions that are both removed and added - ones that are in both tips.

File: src/lib/transaction_pool/transaction_pool.ml (hunk 1 above)
Comment (enolan) [reply src/lib/transaction_pool/transaction_pool.ml]: I made it use `User_command.Stable.Latest.t`.

File: src/lib/transaction_pool/transaction_pool.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bkase) [reply src/lib/transaction_pool/transaction_pool.ml]: oh shit you're right, nevermind

File: src/lib/transaction_pool/transaction_pool.ml [hunk 3]
```diff
@@ -384,7 +390,12 @@ let%test_module _ =
 
     module Test =
       Make0 (struct
-          include Int
+          module Stable = struct
+            module Latest = Int
+          end
+
+          include (Int : module type of Int with module Stable := Int.Stable)
```
Comment (bkase) [inline src/lib/transaction_pool/transaction_pool.ml]: love it

**Pull request merged**